packages/zoe/src/contracts/helpers/bondingCurves.js

```diff
@@ -17,61 +15,82 @@ export const makeConstProductBC = (zoe, issuers) => {
      * several different places, including to check whether an offer is
      * valid, getting the current price for an asset on user request, and
      * to do the actual reallocation after an offer has been made.
-     * @param  {amount[]} poolAmountsArray - an array of the current amount in the
-     * liquidity pool
-     * @param  {amount} amountIn - the amount sent in by a user
+     * @param  {object} poolAmounts - an object of the current amounts in the
+     * liquidity pool keyed by role
+     * @param  {object} amountIn - the amount sent in by a user, keyed
+     * by role
      * @param  {number} feeInTenthOfPercent=3 - the fee taken in tenths of
      * a percent. The default is 0.3%. The fee is taken from amountIn
      */
-    getPrice: (poolAmountsArray, amountIn, feeInTenthOfPercent = 3) => {
+    getPrice: (
```

Comment:
This is a good opportunity for an experiment. Say we replaced `getPrice` with an elaborated form of its internal `getInputPrice`:

```js
    // Calculates how much can be bought by selling input
    const getPrice = ({inputExtent, inputReserve, outputReserve}) => {
      const oneMinusFeeInThousandths = subtract(1000, feeInTenthOfPercent);
      const inputWithFee = multiply(inputExtent, oneMinusFeeInThousandths);
      const numerator = multiply(inputWithFee, outputReserve);
      const denominator = add(multiply(inputReserve, 1000), inputWithFee);

      const outputExtent = floorDivide(numerator, denominator);
      const newOutputReserve = subtract(outputReserve, outputExtent);
      const newInputReserve = add(inputReserve, inputExtent);
      return harden({outputExtent, newOutputReserve, newInputReserve});
    },
```

I suspect this is better than both the old getPrice and this new `getPrice`, or at least not worse. At the point in the callers where you switch to treating role names as data, you could also disaggregate into separate variables, because the roleName object encoding is no longer helping. Neither was the array encoding.

Are there any other contracts where some of their logic treats the role names as data?

Reply:
Thinking about this makes me suspect that the short term plan---where state.js still manipulates arrays, and these are converted back-and-forth to object encodings at the api boundaries---may be the right one. Glad we're doing that first. Later, when we try the pervasive object-roleNames encoding inside, we'll see how that compares. Interesting.

Reply:
I like it! The only issue is that we will have to make amounts out of the extents here, externally, so it's not exactly apples to apples. But I think that's a good division of labor.

Reply:
Ok, this change is in. I've stashed the rest of my autoswap changes and will be making small commits